src/textUtils.c

```diff
@@ -59,39 +59,70 @@ size_t str_formatAdaAmount(uint64_t amount, char* out, size_t outSize)
 	return rawSize + suffixLength;
 }
 
-size_t str_formatUint64(uint64_t number, char* out, size_t outSize)
+static size_t stringifyUint64ToBufferReverse(uint64_t number, char* ptr, char* end)
 {
-	ASSERT(outSize < BUFFER_SIZE_PARANOIA);
-
-	char scratchBuffer[30];
-	char* ptr = BEGIN(scratchBuffer);
-	char* end = END(scratchBuffer);
-
 	// We print in reverse
 	// We want at least one iteration
-	int place = 0;
+	size_t printedChars = 0;
 	do {
 		WRITE_CHAR(ptr, end, '0' + (number % 10));
 		number /= 10;
-		place++;
+		++printedChars;
 	} while (number > 0);
+	return printedChars;
+}
 
-	// Size without terminating character
-	STATIC_ASSERT(sizeof(ptr - scratchBuffer) == sizeof(size_t), "bad size_t size");
-	size_t rawSize = (size_t) (ptr - scratchBuffer);
-
-	if (rawSize + 1 > outSize) {
+static void printReversedStringToBuffer(const char* reverted, size_t revertedSize, char* out, size_t outSize)
+{
+	if (revertedSize + 1 > outSize) {
 		THROW(ERR_DATA_TOO_LARGE);
 	}
-
-	// Copy reversed & append terminator
-	for (size_t i = 0; i < rawSize; i++) {
-		out[i] = scratchBuffer[rawSize - 1 - i];
+	for (size_t i = 0; i < revertedSize; i++) {
+		out[i] = reverted[revertedSize - 1 - i];
 	}
-	out[rawSize] = 0;
+	out[revertedSize] = 0;
+
+	ASSERT(strlen(out) == revertedSize);
+}
 
-	ASSERT(strlen(out) == rawSize);
+size_t str_formatUint64(uint64_t number, char* out, size_t outSize)
+{
+	ASSERT(outSize < BUFFER_SIZE_PARANOIA);
+
+	char scratchBuffer[30];
+	char* ptr = BEGIN(scratchBuffer);
+	char* end = END(scratchBuffer);
+
+	// Size without terminating character
+	const size_t rawSize = stringifyUint64ToBufferReverse(number, ptr, end);
+	printReversedStringToBuffer(scratchBuffer, rawSize, out, outSize);
+	return rawSize;
+}
 
+size_t str_formatInt64(int64_t number, char* out, size_t outSize)
+{
+	ASSERT(outSize < BUFFER_SIZE_PARANOIA);
+
+	char scratchBuffer[30];
+	char* ptr = BEGIN(scratchBuffer);
+	char* end = END(scratchBuffer);
+
+	uint64_t signlessNumber;
+	if (number < 0) {
+		if (INT64_MIN == number) {
+			signlessNumber = ((uint64_t)INT64_MAX) + 1;
+		} else {
+			signlessNumber = (uint64_t) llabs(number);
+		}
+	} else {
+		signlessNumber = (uint64_t) number;
+	}
+	// Size without terminating character (and potential sign)
+	size_t rawSize = stringifyUint64ToBufferReverse(signlessNumber, ptr, end - 1);
+	if (number < 0) {
+		scratchBuffer[rawSize++] = '-';
```

Comment:
Missing `'\0'` at the end of the string? 
I think it is safer to use `snprintf` (take a look at `str_formatAdaAmount` or `bip44_printToStr` for an example).

Reply:
And before calling `printReversedStringToBuffer`, we can assert that rawSize is what it should be, i.e. `strlen(scratchBuffer)`.

Reply:
Rawsize is superfluous if we just count the length again. But sure, I can put a 0 at the end of the scratch area, I just don't see the attack vector here. What _can_ go wrong?

Reply:
A string without \0 is a no-go if it can be avoided. Even if you cannot imagine a scenario where it would go wrong _right now_, sooner or later someone will modify your code or use your function for another purpose, and then in can go wrong easily (e.g. by allowing an attacker to read the memory where some data is stored he should not be able to access).

If a function writes into a string buffer and returns the length of the written string, one should be able to assert that `strlen(buffer) == returned_length`, and it's best to keep this invariant as much as possible even inside functions.

Reply:
Remember, we are somewhat paranoid about security here.